### PR TITLE
Google parser

### PR DESCRIPTION
**Pull request by FelixNeutatz**
This pull request contains:

- Integration of Google Javascript Parser (now we parse all Javascript Strings and search for URLs)
- Minor fixes for URLNormalizer
- Fixing commons dependencies

Metrics:
Execution time for CommonCrawlTest (250MB): 154 s
Output size of Parquet files: 1.2 MB

### CODE REVIEW

File: commons/pom.xml [hunk 1]
```diff
@@ -11,15 +11,27 @@
...
+      <groupId>com.twitter</groupId>
+      <artifactId>parquet-protobuf</artifactId>
+      <version>1.6.0rc1</version>
     </dependency>
+
+
   </dependencies>
 
 
+
```
Comment (sscdotopen) [inline commons/pom.xml]: please remove the empty lines here

File: commons/pom.xml [hunk 2]
```diff
@@ -11,15 +11,27 @@
...
-      <version>2.5.0</version>
+      <version>2.4.0</version>
+    </dependency>
+    -->
+    <dependency>
+      <groupId>com.twitter</groupId>
+      <artifactId>parquet-protobuf</artifactId>
+      <version>1.6.0rc1</version>
     </dependency>
+
```
Comment (sscdotopen) [inline commons/pom.xml]: please remove the empty lines here

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 3]
```diff
@@ -16,161 +16,288 @@
...
+import java.util.HashSet;
+import java.util.Set;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
 
-import com.google.common.collect.Sets;
 
 public class ResourceExtractor {
 
-  private static final Logger LOG = LoggerFactory.getLogger(ResourceExtractor.class);
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: please switch back to two space indentation

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 4]
```diff
@@ -16,161 +16,288 @@
...
+
+    private static Set<String> EXTRA_ANNOTATIONS = new HashSet<String>(Arrays.asList(
+            "suppressReceiverCheck",
+            "suppressGlobalPropertiesCheck"
+    ));
+
+
+    public synchronized Iterable<Resource> extractResources(String sourceUrl, String html) {
+
+        ArrayList<String> scriptHtml = new ArrayList<String>();
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: scriptHtml should be of generic type List

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 5]
```diff
@@ -16,161 +16,288 @@
...
+        String prefixForInternalLinks = urlNormalizer.createPrefixForInternalLinks(sourceUrl);
+
+        Document doc = Jsoup.parse(html);
+        Elements iframes = doc.select("iframe[src]");
+        Elements links = doc.select("link[href]");
+        Elements imgs = doc.select("img[src]");
+        Elements scripts = doc.select("script");
 
-  private final URLNormalizer urlNormalizer = new URLNormalizer();
+        Elements all = iframes.clone();
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: please rename all to allElements

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 6]
```diff
@@ -16,161 +16,288 @@
...
+        all.addAll(scripts);
+        all.addAll(links);
+        all.addAll(imgs);
 
-  private final Pattern javascriptPattern = Pattern.compile("((\"|\')(([-a-zA-Z0-9+&@#/%?=~_|!:,;\\.])*)(\"|\'))");
+        String uri;
 
-  private static final int STACK_OVERFLOW_LIMIT = 6000;
+        for (Element tag: all) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space before : please

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 7]
```diff
@@ -16,161 +16,288 @@
...
+                    }
+                }
+                if (isValidDomain(uri)) {
+                    resources.add(new Resource(uri, type(tag.tag().toString())));
+                }
+            }
+
+
+            if (tag.tag().toString().equals("script")) { //filter functions
+                if(tag.data().length() > 1) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space after if please

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 8]
```diff
@@ -16,161 +16,288 @@
...
+                    scriptHtml.add(tag.data());
+                }
+            }
+        }
+
+        Config config = ParserRunner.createConfig(
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: no line break here, please

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 9]
```diff
@@ -16,161 +16,288 @@
...
+            public void error(String message, String sourceName, int line, int lineOffset) {
+                if (LOG.isWarnEnabled()) {
+                    //LOG.warn("Parser Error: \"" + message + "\"");
+                }
+            }
+        };
+
+        StaticSourceFile f = new SimpleSourceFile("input", false);
+
+        ArrayList<String> parsedStrings = new ArrayList<String>();
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: can be of generic type list

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 10]
```diff
@@ -16,161 +16,288 @@
...
+
+        for(String script:scriptHtml) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space after for and around : please

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 11]
```diff
@@ -16,161 +16,288 @@
...
+                ArrayList<String> variables = new ArrayList<String>();
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: can be of generic type List

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 12]
```diff
@@ -16,161 +16,288 @@
...
+                ArrayList<String> docVariables = new ArrayList<String>();
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: can be of generic type List

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 13]
```diff
@@ -16,161 +16,288 @@
...
+                try {
+                    ParserRunner.ParseResult r = ParserRunner.parse(f, script, config, errorReporter);
+                    //printTree(r.ast, 0);
+
+                    parseStrings(r.ast, parsedStrings);
+                } catch(Exception e) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space after catch please

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 14]
```diff
@@ -16,161 +16,288 @@
...
-            for(int i = 0; i < matcher.groupCount(); i++) {
-              String url = matcher.group(i);
-
-              if (url != null && url.contains(".") && !url.contains("\"") && !url.contains("'")) {
-                if (!url.contains(":") || url.contains("://")) {
-                  try {
+    //parse url when it is within the string
+    private void tokenizeStrings(ArrayList<String> parsedStrings) {
+
+        ArrayList<Integer> toBeReplaced = new ArrayList<Integer>();
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: can be of generic type List

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 15]
```diff
@@ -16,161 +16,288 @@
...
+
+        ArrayList<String> tokenizedStrings = new ArrayList<String>();
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: can be of generic type List

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 16]
```diff
@@ -16,161 +16,288 @@
...
+
+        for(int o=0;o<parsedStrings.size();o++) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: spaces please

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 17]
```diff
@@ -16,161 +16,288 @@
...
+            String currentString = parsedStrings.get(o);
+
+            if( currentString.contains("\"") || currentString.contains("'")) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space before ( none after

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 18]
```diff
@@ -16,161 +16,288 @@
...
+                Matcher matcher = javascriptPattern.matcher("'" + currentString + "'");
+                boolean found = false;
+                while (matcher.find()) {
+                    if (found == false) {
+                        found = true;
+                        toBeReplaced.add(o);
+                    }
+
+                    for(int i = 0; i < matcher.groupCount(); i++) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space after for

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 19]
```diff
@@ -16,161 +16,288 @@
...
+
+                        if (token != null && !token.contains("\"") && !token.contains("'") && isUrl(token)) {
+                            tokenizedStrings.add(token);
+                        }
+                    }
+                }
+            }
+        }
+
+        for(int o=toBeReplaced.size()-1;o>=0;o--) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: spaces :)

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 20]
```diff
@@ -16,161 +16,288 @@
...
 
-    return resources;
-  }
+    private boolean isUrl(String url) {
 
-  private boolean isValidDomain(String url) {
-    if (!url.contains(".") || url.contains("///")) {
-      return false;
-    }
+        if(!url.contains(".")) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space after if

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 21]
```diff
@@ -16,161 +16,288 @@
...
+            return false;
+        }
 
-    if (url.contains(";") || url.contains("=") || url.contains("?")) {
-      return false;
-    }
+        //remove and check white space
+        url = url.trim();
+        if(url.contains(" ") || url.contains("\t") || url.contains("\r") || url.contains("\n")) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space after if

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 22]
```diff
@@ -16,161 +16,288 @@
...
+            return false;
+        }
 
-    int startTopLevelDomain = url.lastIndexOf('.');
-    String topLevelDomain = url.substring(startTopLevelDomain + 1);
-    return DomainValidator.getInstance().isValidTld(topLevelDomain);
-  }
+        //TODO: check this condition
+        //this doesnt work for something like localhost:80/...
+        if(url.contains(":")) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space after if

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 23]
```diff
@@ -16,161 +16,288 @@
...
-  private Resource.Type type(String tag) {
-    if ("script".equals(tag)) {
-      return Resource.Type.SCRIPT;
+        return true;
     }
-    if ("link".equals(tag)) {
-      return Resource.Type.LINK;
+
+    private void printTree(Node root, int level) {
+        for(int i = 0; i < level; i++) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space after for

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 24]
```diff
@@ -16,161 +16,288 @@
...
+
+        for(Node child : root.children()){
+            printTree(child,level+1);
+        }
     }
-    if ("img".equals(tag)) {
-      return Resource.Type.IMAGE;
+
+    private void parseStrings(Node root, ArrayList<String> parsedStrings) {
+        if(root.isString()) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space after if

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java [hunk 25]
```diff
@@ -16,161 +16,288 @@
...
+            if(root.getString().contains(".")) {
+                parsedStrings.add(root.getString());
+            }
+        }
+
+        for(Node child : root.children()){
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/ResourceExtractor.java]: space after for

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLNormalizer.java [hunk 26]
```diff
@@ -127,8 +127,21 @@ public String createPrefixForInternalLinks(String sourceUrl) {
     return prefixForInternalLinks;
   }
 
+    private String deletePHP(String url) {
+        if(url.contains("?")) {
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLNormalizer.java]: space after if

File: extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLNormalizer.java [hunk 27]
```diff
@@ -127,8 +127,21 @@ public String createPrefixForInternalLinks(String sourceUrl) {
...
+
     public String normalize(String urlString)
             throws MalformedURLException {
+
+        urlString = urlString.trim();
+
+        urlString = deletePHP(urlString);
+
+
         if ("".equals(urlString))                     // permit empty
```
Comment (sscdotopen) [inline extraction/src/main/java/io/ssc/trackthetrackers/extraction/resources/URLNormalizer.java]: please always use braces with if

**Comment by sscdotopen** [comment]
I mostly have style comments, could you please fix them? Afterwards I will merge the PR and test it.

We use the standard Sun code conventions for Java with 2 spaces indent and 120 characters per line :)

**Comment by FelixNeutatz** [comment]
done :)

**Pull request merged**